### PR TITLE
Add developer responsiveness badges

### LINKED ISSUE #757: High response rate badge
Another way to help developers stand out on the platform is to highlight their response rate. Businesses are more inclined to reach out to someone who responds to every message than someone who doesn't.

I'm defining "response rate" as the number of conversations a developer has responded (at all) divided by the total number of conversations a developer is participating in.

A PR addressing this issue should:

- [ ] Add a badge to developer profiles if their response rate is over 90%
- [ ] Replace that badge with a different one if their response rate is a perfect 100%
- [ ] If needed, ensure the badge status is cached efficiently with no N+1 queries
- [ ] Stretch goal: Give developers 6 hours to respond to a new message before _dropping or removing_ their badge

For the badges, please use the following design/copy for a first pass.

* `90% response rate` in indigo
* `100% response rate` in purple

### PR DESCRIPTION
**Pull request by fig**
This PR adds a "responsiveness" badge to a developer's profile.
closes #757

## Pull request checklist

- [x] My code contains tests covering the code I modified
- [x] I linted and tested the project with `bin/check`
- [x] I added significant changes and product updates to the [changelog](CHANGELOG.md)

## Requirements from #757 
 - [x] Add a badge to developer profiles if their response rate is over 90%
 - [x] Replace that badge with a different one if their response rate is a perfect 100%
 - [x] If needed, ensure the badge status is cached efficiently with no N+1 queries
       Requires review by someone with better database fu.
 - [x] Stretch goal: Give developers 6 hours to respond to a new message before dropping or removing their badge
        Too follow

I'm not totally happy with the text used in the badges so any suggestions for alternatives would be gratefully appreciated.
Current text: "Good responder" / "Perfect responder"

I have chosen to implement this feature by writing a concern to be included in the `Developer` model. The concern utilises the existing `Stats::Conversation` class to calculate developer response rate.

## Sample rendering

image

### CODE REVIEW

**Comment by joemasilotti** [comment]
Thanks for the submission!

Let's use the badge titles I outlined in the issue for now: "100% response rate" and "90% response rate".

Also, I won't be able to merge this until the third todo item is completed. Without caching the values we end up with a nasty N+1 on `/developers`:

```
web      |   Conversation Count (0.3ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 26]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.3ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 25]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.3ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 20]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.3ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 19]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.3ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 18]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.4ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 17]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.4ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 16]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.3ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 15]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
web      |   Conversation Count (0.3ms)  SELECT COUNT(*) FROM "conversations" WHERE "conversations"."developer_id" = $1 AND "conversations"."developer_blocked_at" IS NULL AND "conversations"."business_blocked_at" IS NULL  [["developer_id", 14]]
web      |   ↳ app/models/stats/conversation.rb:10:in `sent'
```

Let me know if you have appetite to take care of that. If no, no worries! I can take over.

**Comment by fig** [comment]
All noted.

After sleeping on this, I'm not loving the dependency on `Stats::Conversation` during the request cycle. Maybe adding a database column for `response_rate` might be a better way to go architecturally. I envision a background job queued to update this column a configurable time (e.g. 12 hours) after a 'cold message' is received.

This would solve the N+1 and keep the use of `Stats::Conversation` to a minimum and in a background job. Essentially, we will cache the `response_rate` in the db.

Your thoughts?

**Comment by joemasilotti** [comment]
That's very similar to what I was thinking, too!

We can roll ahead and cache this directly in the `developers` table for now. Then when #758 is implemented we can migrate over the data.

Or, we can hold off and wait until that is merged in first. Your call!

**Comment by joemasilotti** [comment]
It will generate on additional db query per _list_ of developers, instead of one for every badge for every developer.

So, for each paginated page of 10 developers it will add 1 additional query.

If we had all of the current and upcoming badge logic calculated on the fly it would generate 40 additional queries (4 badges x 10 developers).

**Comment by fig** [comment]
- The solution now adds a new column `response_rate` to the `developers` table.
- A newly created conversation triggers a background job to update the Developer's response rate after a configurable length of time.
- Badge text fixed.

A migration is included in this PR
A rake task is provided to backfill the new table column (`rake backfills:developer_response_rate`).

File: app/controllers/cold_messages_controller.rb
Comment (joemasilotti) [inline app/controllers/cold_messages_controller.rb:62]: This feels like a round about way to queue the job. Let's instead add this to `Messages::Notifications` or inline to the controller.

File: lib/tasks/backfills.rake
Comment (joemasilotti) [inline lib/tasks/backfills.rake:9]: Let's remove this and let the job fail on its own when the migration hasn't been run.

File: lib/tasks/backfills.rake
Comment (joemasilotti) [inline lib/tasks/backfills.rake:14]: Why sleep?

File: app/jobs/update_developer_response_rate_job.rb
Comment (joemasilotti) [inline app/jobs/update_developer_response_rate_job.rb:9]: I don't think this logic is correct. Isn't it only querying for conversations started within the last 12 hours?

Instead we need something that _ignores_ all conversations created in the last 12 hours and calculates the response rate of the others.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
Looking good!

image

I'm now realizing that 90 vs. 100 might not matter very much. Let's consolidate them. If a developer has a 90%+ response rate show a badge that says "High response rate".

Also, it looks like the background job is updating the developer's `updated_at` column which is used to show the "Recently active" badge. Let's update the job to not update that column for now with a comment explaining why. This will be removed in a future PR when we start to cache the badges in their own model.

I also made some inline questions/comments.

File: app/controllers/cold_messages_controller.rb
Comment (fig) [reply app/controllers/cold_messages_controller.rb]: I didn't like it here either, but it didn't seem to fit in with the Notifications either. It possibly deserves it's own class?
I've moved it in with the Notifications for now, as suggested.

File: lib/tasks/backfills.rake
Comment (fig) [reply lib/tasks/backfills.rake:6]: Sleep to throttle mass Database insertions between batches of 50.
I'm not a db expert, so I don't totally understand the reasoning or the choice of value, but I've read that this is a GoodIdea™.
Refs:
https://world.hey.com/paulosilva/ruby-on-rails-backfilling-data-136adb66
https://github.com/ankane/strong_migrations#backfilling-data

File: app/jobs/update_developer_response_rate_job.rb
Comment (fig) [reply app/jobs/update_developer_response_rate_job.rb]: `developer.conversations.where("created_at < ?", GRACE_PERIOD.ago)`
Fetch conversations where `created_at` is less than (i.e earlier than) the given time value.

I've added tests to confirm the logic.

**Comment by fig** [comment]
I've changed the implementation to have a single "High Response Rate" badge as suggested.

I noticed that "recently active" was relying on the `updated_at` column. The background job is calling `#update_column` which updates the attribute directly in the database issuing an UPDATE SQL statement. I chose this method because updated_at/updated_on are not updated. One downside is that validations are also skipped. I've add a comment to change this to a regular `#update!` at a later date.

I noticed that `test/components/developers/badge_component_test.rb` wasn't actually doing what it claimed.
It was possible to delete the lines setting the attributes of the developer and the tests would still pass.
I have renamed the test file to `test/components/developers/badges_component_test.rb` and refactored the tests within to exercise the Developers::BadgesComponent. This class utilises the BadgeComponent for rendering and so now the whole chain is tested.

File: lib/tasks/backfills.rake
Comment (joemasilotti) [reply lib/tasks/backfills.rake]: Got it. TIL, thanks!

File: app/jobs/update_developer_response_rate_job.rb
Comment (joemasilotti) [reply app/jobs/update_developer_response_rate_job.rb]: Ah! You're right. I read that backwards. Thanks for clarifying - and with tests to prove it.

**Comment by joemasilotti** [comment]
What's even weirder is that when I run `bin/rails rake db:reset` in development I get the same error. But when starting a new conversation from the UI the job enqueues without issue.

Honestly, I'm not really sure what's going on! I'll have to dig in more.

**Comment by fig** [comment]
This is what I have learnt in the last few hours.

- Unless specified otherwise, the data seeding task will use the :async queue adaptor.
- We are specifying the :inline adaptor in our seeds file, so this will be honoured.
- The :inline adaptor does not allow scheduling jobs so raises an error when trying to seed Conversations because of the `.set(wait: ...)` method.
- If we specify :async only for the UpdateDeveloperResponseRateJob,
  (`UpdateDeveloperResponseRateJob.queue_adapter = :async` in seeds.rb), the db will seed, but will enqueue jobs in our dev environment.

**Comment by joemasilotti** [comment]
Got it! Thanks for identifying all the moving parts - that's helpful.

Ignoring the implementation details for a second... when running the seeds ideally the column is populated right away. Meaning, the job kicks off immediately.

Is it possible to achieve that? Perhaps via running the job conditionally async/sync depending on the environment? Or something more graceful in the config or options passed to the job?

**Comment by fig** [comment]
One option might be to set  `config.developer_response_grace_period` to a very low value (5secs) in development and have the job run :async during seeds. The low grace period won't be an issue as our seeded developers will be responding almost instantaneously and the update job will run as soon as a dev server is fired up. My only concern is that there will be jobs left dormant in redis if a server instance is not started after seeding. After saying that, I believe the response rate update job is idempotentish (multiple calls within a short time-frame will produce the same result) so it might not be an issue.

File: db/seeds/conversations.rb
Comment (joemasilotti) [inline db/seeds/conversations.rb:7]: I'm confused as to why this needs to be performed outside of the normal creating of messages. Should the job get kicked off "in 0 seconds" when running the seeds?

File: test/jobs/update_developer_response_rate_job_test.rb
Comment (joemasilotti) [inline test/jobs/update_developer_response_rate_job_test.rb:35]: Why a song and dance with the mock instead of checking for `50`? We clearly create both a replied to and not replied to conversation in the setup of this test, so no risk for mystery guests.

File: test/jobs/update_developer_response_rate_job_test.rb
Comment (joemasilotti) [inline test/jobs/update_developer_response_rate_job_test.rb:45]: We can use the following for this test.

```ruby
assert_no_changes @developer.reload.response_rate do
  # ...
end
```

File: test/jobs/update_developer_response_rate_job_test.rb
Comment (joemasilotti) [inline test/jobs/update_developer_response_rate_job_test.rb:64]: The relative assertions are a little hard to reason about. Let's we instead use `assert_changes @developer.reload.response_rate, from: X, to: Y`. I think that's a bit clearer.

**Comment by fig** [comment]
- Set the response grace period to zero in production and use the :async adaptor to avoid the seeding issue.
- Added a call to the update job after seeding conversations — this ensures that a freshly seeded development environment starts with appropriate developer badges.
- Fixed an issue with the response rate logic whereby conversations replied to within the grace period were not being taken into account.

The test files (especially [update_developer_response_rate_job_test.rb](https://github.com/joemasilotti/railsdevs.com/pull/774/files#diff-0a8de7757c0bf4b2a5663c45cfa5cc208860c0440eff73d64949d1f9b9bc2ad9) ) are probably not very pretty. I usually use RSpec, so feel free to refactor them, or open an issue to get them tidied up. I believe their functionality is sound but, again, they could probably use a closer look by someone proficient in MiniTest.

File: CHANGELOG.md
Comment (joemasilotti) [inline CHANGELOG.md:7]:
```suggestion
* February 28 - Add a badge for developers with a high response rate #774 @fig
```

**Review by joemasilotti — CHANGES_REQUESTED** [review]
Great idea using `0.seconds`! Just a few more minor requests and this is ready to merge.

File: app/models/conversation.rb
Comment (fig) [reply app/models/conversation.rb:72]: Ah yes! Good spot.

File: db/seeds/conversations.rb
Comment (fig) [reply db/seeds/conversations.rb:7]: I thought it would, but it didn't appear to be doing so.
I shall conduct further experiments.

File: test/jobs/update_developer_response_rate_job_test.rb
Comment (fig) [reply test/jobs/update_developer_response_rate_job_test.rb:35]: This is my brain trying to translate RSpec to MiniTest. 🙃
I'll see if I can make it clearer.

File: db/seeds/conversations.rb
Comment (fig) [reply db/seeds/conversations.rb]: Working as expected now without the call to the job.

File: app/models/conversation.rb
Comment (fig) [reply app/models/conversation.rb:72]: Test added for 'Conversation#developer_replied?'

File: test/jobs/update_developer_response_rate_job_test.rb
Comment (fig) [reply test/jobs/update_developer_response_rate_job_test.rb]: I have done an almost complete rewrite of the `test/jobs/update_developer_response_rate_job_test.rb` 
I've learn't a lot. I see more MiniTest in my future! :crystal_ball:

File: test/jobs/update_developer_response_rate_job_test.rb
Comment (fig) [reply test/jobs/update_developer_response_rate_job_test.rb]: TIL! Thank you.
Done.

File: test/jobs/update_developer_response_rate_job_test.rb
Comment (fig) [reply test/jobs/update_developer_response_rate_job_test.rb]: Another thing I didn't know until today. Thank you.

**Review by joemasilotti — APPROVED** [review]
Looks great! Nice work @fig. Thanks for all the updates.

**Pull request merged**
